msglist: Fetch more messages despite previous muted batch

sm-sayedi opened the pull request:
Fixes: #1256

<details>
<summary>Screen recordings</summary>

### First batch - Empty

#### Before
https://github.com/user-attachments/assets/9d631020-262e-4fed-94b5-9c40485d3787

#### After
https://github.com/user-attachments/assets/de47ac15-3c15-48f1-a271-91d0604ab1ff

### First batch - Less than a screenful

#### Before
https://github.com/user-attachments/assets/6db2189f-5486-497c-bc0b-4642a12dafb1

#### After
https://github.com/user-attachments/assets/30582988-4deb-466a-81e1-b4add904c2a1

</details>

sm-sayedi committed:
msglist [nfc]: Mention fetchNewer in the MessageListView dartdoc

sm-sayedi committed:
msglist: Fetch more messages despite previous muted batch

Fixes #1256.

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -127,6 +127,10 @@ mixin _MessageSequence {
   ///
+  /// This also may or may not represent all the message history that
+  /// conceptually belongs in this narrow because some messages might be
+  /// muted in one way or another and they may not appear in the message list.
```
What's the distinction between "conceptually belongs in this message list" and "conceptually belongs in this narrow"? 

This paragraph feels very focused on the distinction between  "conceptually belongs in this message list" and "conceptually belongs in this narrow", because of its parallelism with the previous paragraph:

```dart
  /// This may or may not represent all the message history that
  /// conceptually belongs in this message list.
  /// That information is expressed in [fetched], [haveOldest], [haveNewest].
```

Is that focus intentional?

There's another place we draw a similar-looking distinction, in `Narrow.containsMessage`:

```dart
  /// […]
  ///
  /// This does not necessarily mean the message list would show this message
  /// when navigated to this narrow; in particular it does not address the
  /// question of whether the stream or topic, or the sending user, is muted.
  /// 
  /// […]
  bool? containsMessage(MessageBase message);
```

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -937,23 +937,20 @@ class MessageListView with ChangeNotifier, _MessageSequence {
 
       processResult(result);
+    } catch (e) {
+      hasFetchError = true;
```
```
msglist [nfc]: Remove one nested try block in `_fetchMore`

The nested `try` block doesn't seem to be making any difference,
so good to remove.
```

This isn't NFC: now, `hasFetchError` is about what happens when data is fetched _and processed_ (with `processResult`), not just about what happens when data is fetched. That doesn't seem desirable, just from looking at its name.

chrisbobbe reviewed (commented):
Thanks! Here's a review of the first three commits:

542996e5d msglist [nfc]: Mention fetchNewer in the MessageListView dartdoc
ce5645636 msglist [nfc]: Add a new point to _MessageSequence.messages dartdoc
3c10bedb1 msglist [nfc]: Remove one nested `try` block in `_fetchMore`

and a partial review of the fourth:

03d61ca54 msglist: Fetch newer messages despite previous muted batch

For that fourth commit, can you say briefly in the commit message why it's not a complete fix for the issue, to help orient the reader to what comes next?

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -1222,12 +1217,13 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
         // TODO perhaps offer mark-as-read even when not done fetching?
-        MarkAsReadWidget(narrow: widget.narrow),
+        if (model.messages.isNotEmpty)
+          MarkAsReadWidget(narrow: widget.narrow),
```
What's the relationship of this change to the main changes happening in this commit? What's the user-visible change in behavior it causes?

sm-sayedi replied:
So with the new changes in the commit, if the initial newest batch is all muted (`model.messages.isEmpty`), and then when older messages are being fetched, the "Mark as read" widget will be shown along with a progress indicator and no messages, so I thought it may be a good idea to hide "Mark as read" until there are visible messages populated.
| Before (this one-line change) | After ((this one-line change)|
| --- | --- |
| <img width="546" height="1040" alt="Screenshot 2025-11-26 at 7 30 45 AM" src="https://github.com/user-attachments/assets/a8ee3124-f9e3-4cfa-bf6c-f4c8d78ddec2" /> | <img width="546" height="1040" alt="Screenshot 2025-11-26 at 7 32 13 AM" src="https://github.com/user-attachments/assets/2338585c-a50b-4dac-9c19-56491a958c07" /> |

sm-sayedi replied:
Yeah, I think that will be an improvement. It will also avoid the abrupt change from a centered progress indicator to a bottom-aligned indicator.

gnprice commented on lib/model/message_list.dart:
```diff
@@ -163,19 +201,34 @@ mixin _MessageSequence {
+
+  FetchingInitialStatus _fetchInitialStatus = .unstarted;
+  FetchingMoreStatus _fetchOlderStatus = .unstarted;
+  FetchingMoreStatus _fetchNewerStatus = .unstarted;
```
More broadly:

    But imagine if the number of messages in the initial batch occupies less
    than a screenful, and then `fetchOlder` returns no messages or a few
    messages that combined with the initial batch messages are still less
    than a screenful; in that case, there will be no change in the scroll
    metrics to call `fetchNewer`.

This change feels like the wrong layer for solving this problem.

When the `fetchOlder` returns, that will notify the view-model's listeners, right? What if we have the _MessageListState react to that by looking to see if more fetching is needed, and calling `fetchOlder`/`fetchNewer` if so?

IOW, we could have its `_modelChanged` method call the same logic that's at the end of `_handleScrollMetrics`. Probably that means pulling that part out as its own smaller helper method.

gnprice reviewed (commented):
Thanks!

Similar to https://github.com/zulip/zulip-flutter/pull/1951#pullrequestreview-3507432445, here's some high-level comments before I go on vacation.

sm-sayedi committed:
msglist test [nfc]: Move a few helpers up to the top, for reusability

These are: itemCount, findPlaceholder, and findLoadingIndicator.

Also, rename itemCount to messageListItemCount, to make it clearer
what it refers to in the upcoming tests.

sm-sayedi commented:
Thanks @chrisbobbe and @gnprice for the previous reviews. Pushed a new revision, PTAL.

sm-sayedi commented on test/widgets/message_list_test.dart:
```diff
@@ -872,6 +885,180 @@ void main() {
+      check(messageListItemCount(tester)).equals(2 + 2 + 98);
+    });
+
+    testWidgets('mid-history and fetch-older with too few messages, fetch-newer request is made', (tester) async {
```
This test is for handling an edge case, the handling of which required a lot of changes in the first revision of this PR, before [Greg suggested](https://github.com/zulip/zulip-flutter/pull/1989#discussion_r2562220713) an alternative, straightforward method of solving that problem. I explained that edge case in the description of a commit message, which is not included now.
```
This change is necessary when there is a need to fetch more messages
in both directions, older and newer, and when fetching in
one direction avoids fetching in another direction at the same time,
because of the `if (busyFetchingMore) return` line in
both `fetchOlder` and `fetchNewer`.

This scenario happens when a conversation is opened in its first unread,
such that the number of messages in the initial batch is so low (because
they're muted in one way or another) that it's already past the certain
point where the scroll metrics listener in the widget code triggers both
`fetchOlder` and `fetchNewer`. As of 2025-11, that code first calls
`fetchOlder` then `fetchNewer`, and for the reason mentioned above,
`fetchNewer` will not fetch any new messages. But that's fine, because
as soon as older messages from `fetchOlder` arrives, there will be
a change in the scroll metrics, so `fetchNewer` will be called again,
fetching new messages.

But imagine if the number of messages in the initial batch occupies less
than a screenful, and then `fetchOlder` returns no messages or a few
messages that combined with the initial batch messages are still less
than a screenful; in that case, there will be no change in the scroll
metrics to call `fetchNewer`.

With the three fetch request types being separated, especially the two
request types for older and newer messages, each direction can fetch
more messages independently without interfering with one another.
```

chrisbobbe reviewed (commented):
Thanks! Comments below, from reading everything except the tests.

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -868,6 +903,12 @@ class MessageListView with ChangeNotifier, _MessageSequence {
+      await fetchOlder(partOfInitialFetch: true);
+      if (messages.isNotEmpty) break;
+      await fetchNewer(partOfInitialFetch: true);
+    }
```
Ah yeah you're right, that makes sense.

chrisbobbe reviewed (commented):
Thanks! A few comments below, including one that should help reduce complexity. 🙂

sm-sayedi commented on lib/model/message_list.dart:
```diff
@@ -811,9 +841,14 @@ class MessageListView with ChangeNotifier, _MessageSequence {
+  /// If the results don't include at least one non-muted message,
+  /// this will call [fetchOlder] and/or [fetchNewer]
+  /// until one is found or the narrow's oldest and newest messages are reached.
   Future<void> fetchInitial() async {
```
Yeah, removing the loop inside `fetchInitial` will certainly simplify the code, but it will have some effects on the UI, for example, if the initial batch has no visible messages, the loading indicator in the middle will suddenly move to the bottom. This change was suggested in https://github.com/zulip/zulip-flutter/pull/1989#discussion_r2566446114 with the initial motivation for it mentioned in there.

sm-sayedi committed:
msglist: Autofocus compose box only when the whole history is empty

Previously, if the initial batch came up empty, then the history was
considered empty too. But now, the initial batch could be empty
because of the messages in muted conversations while there may still be
other messages in history.

chrisbobbe commented:
Great, thanks! LGTM; marking for Greg's review.

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -812,33 +812,36 @@ void main() {
-      await tester.pump(const Duration(milliseconds: 1));
+      // Allow a delayed frame for the response of the first batch to arrive.
+      await tester.pump(Duration(milliseconds: 1));
+      check(itemCount(tester)).equals(401);
```
Let's also insert a check that verifies at what step here the second fetch request gets made.

gnprice replied:
Also, the story here seems a bit ambiguous:
* The comment at the start of the range I've highlighted says the second fetch is "on the next frame" after the first fetch starts.
* But then the logic at the end of this range seems to be that we wait all the way until the first fetch *finishes*. This happens to also be the first frame after it starts.

If there are several frames between the first fetch starting and ending (which is going to be the typical case — most of these fetches take much longer than a millisecond), then those two descriptions give different predictions for when the second fetch happens. Which one is the behavior that occurs?

sm-sayedi replied:
By “on the next frame” I meant the next frame in the test code (which is the delayed pump after which the first batch arrives). And in real scenarios, the second fetch starts just after the first one finishes, not after the first frame post first fetch start. Changed that comment to reflect this fact.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -1032,17 +1063,17 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
-      model.fetchNewer();
+    if (SchedulerBinding.instance.schedulerPhase == .transientCallbacks) {
+      SchedulerBinding.instance.addPostFrameCallback((_) {
+        if (scrollController.hasClients) {
```
What's the reasoning for this condition?

sm-sayedi replied:
In this case I think it doesn't really make any difference, but it's a safety check. Maybe there is a possibility that `scrollController` can lose clients post frame; I'm not sure. The same check is necessary for another `addPostFrameCallback` in one of the next commits, though, so I thought it would be a good idea to add it here too.

gnprice reviewed (commented):
Thanks @sm-sayedi for fixing these issues, and @chrisbobbe for the previous reviews!

Comments below up through fixing the double-fetch glitch #2104:
9f0f456ce msglist [nfc]: Mention fetchNewer in the MessageListView dartdoc
25886d90c msglist [nfc]: Add a new point to _MessageSequence.messages dartdoc
ef20b6d2b msglist [nfc]: Point the TODO about the double-fetch glitch to an issue
bb83ec0d5 msglist test: Reproduce the double-fetch glitch
3c4cd603e msglist [nfc]: Extract the "fetch more messages" code in a new method
13678fe68 msglist: Fix the double-fetch glitch

Let's split out the #2104 part to its own separate PR. It feels like several different things are happening in this PR, and splitting it up will help with managing review. I believe that means these commits:
ef20b6d2b msglist [nfc]: Point the TODO about the double-fetch glitch to an issue
bb83ec0d5 msglist test: Reproduce the double-fetch glitch
3c4cd603e msglist [nfc]: Extract the "fetch more messages" code in a new method
13678fe68 msglist: Fix the double-fetch glitch

sm-sayedi commented:
Thanks @gnprice for the review. Revision pushed, PTAL.

Opened a new PR (#2162) for those four commits about fixing the double-fetch glitch and rebased this PR atop that.

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -945,14 +945,14 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
-      // If the fetch came up empty, there's nothing to read,
-      // so opening the keyboard won't be bothersome and could be helpful.
+    if (model.messages.isEmpty && model.haveNewest && model.haveOldest && !_hasAutofocused) {
+      // If there are no messages to show in the whole history,
+      // opening the keyboard won't be bothersome and could be helpful.
       // It's definitely helpful if we got here from the new-DM page.
       MessageListPage.ancestorOf(context)
         .composeBoxState?.controller.requestFocusIfUnfocused();
+      _hasAutofocused = true;
     }
-    _prevFetched = model.fetched;
```
Oh, Claude made a good point that I missed in my suggestion involving `_hasAutofocused` here. With this code, the autofocus could end up firing at an unwanted time:

1. The initial message fetch returns at least one message. (Autofocus doesn't happen now.)
2. Later, the message list empties out, by messages being deleted, leaving `haveNewest` and `haveOldest` both true. But `_hasAutofocused` is still false, because autofocus wasn't triggered in (1).
3. Autofocus triggers, bringing up the keyboard and distracting the user.

Instead, I think we'd want to gate on whether the decision has been made at all, rather than on whether we decided to autofocus. Claude suggests something like:

```dart
    if (!_autofocusDecided && model.fetched
        && (model.messages.isNotEmpty || (model.haveOldest && model.haveNewest))) {
      _autofocusDecided = true;
      if (model.messages.isEmpty) {
        // If there are no messages to show in the whole history, […]
        MessageListPage.ancestorOf(context)
          .composeBoxState?.controller.requestFocusIfUnfocused();
      }
    }
```

What do you think? Would you also write some tests for this behavior?

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -911,12 +944,13 @@ class MessageListView with ChangeNotifier, _MessageSequence {
+    assert(oldestFetchedMessageId != null);
     await _fetchMore(
-      anchor: NumericAnchor(messages[0].id),
+      anchor: NumericAnchor(oldestFetchedMessageId!),
```
Claude is flagging some gaps in test coverage; would you take a look? All the tests in test/model/message_list_test.dart still pass if I revert this change; likewise for the similar line in `fetchNewer`. Would you iterate with Claude to make sure we've got all the essential test coverage, including the cases where `oldestFetchedMessageId` / `newestFetchedMessageId` differ from `messages[0].id` / `messages.last.id`?

chrisbobbe reviewed (commented):
Thanks! Comments below.

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -938,12 +971,13 @@ class MessageListView with ChangeNotifier, _MessageSequence {
       numBefore: kMessageListFetchBatchSize,
       numAfter: 0,
       processResult: (result) {
+        _oldestFetchedMessageId = result.messages.firstOrNull?.id ?? _oldestFetchedMessageId;
```
nit: let's give this long line some breathing room, e.g.:

```suggestion
        if (result.messages.isNotEmpty) {
          _oldestFetchedMessageId = result.messages.first.id;
        }
```

chrisbobbe commented on lib/model/message_list.dart:
```diff
@@ -968,12 +1002,13 @@ class MessageListView with ChangeNotifier, _MessageSequence {
       numBefore: 0,
       numAfter: kMessageListFetchBatchSize,
       processResult: (result) {
+        _newestFetchedMessageId = result.messages.lastOrNull?.id ?? _newestFetchedMessageId;
```
(Similarly here.)

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -894,7 +894,7 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
   }
 
-  bool _prevFetched = false;
+  bool _autofocusDecided = false;
```
```
msglist: Autofocus compose box only when the whole history is empty

Previously, if the initial batch came up empty, then the history was
considered empty too. But now, the initial batch could be empty
because of the messages in muted conversations while there may still be
other messages in history.
```

Commit-message nit:

> the initial batch could be empty because of the messages in muted conversations

This part was already true before the PR; the change here is that now we keep fetching before deciding to autofocus.
